### PR TITLE
Prevent broken concurrent schema read/writes

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

**Comment by bereng** [comment]
@maedhroz the 3.11 version is slightly different. Both classes are in different packages, also `SchemaKeyspace` is already sync'ing on static on a few methods. It makes little sense to me to move everything into `Schema` imo. The easiest is to add static sync'ing where needed.

I have trouble making your test fail despite having tried several hook points for the BM. I could use a little help here. Edit: as usual next day you find the problem in 5m. I had to make the sleep 5s but it certainly fails now when sync is removed.

**Comment by bereng** [comment]
I have to look into this one, it is deadlocking now https://app.circleci.com/pipelines/github/bereng/cassandra/476/workflows/970669eb-e533-42b8-a9ef-d96ff2fcf04d/jobs/4135/parallel-runs/0/steps/0-104

Edit: Had forgotten an extra sync

**Comment by bereng** [comment]
CI [lgtm](https://app.circleci.com/pipelines/github/bereng/cassandra/478/workflows/df553f68-8121-43e5-a1d5-2e49af6ec270)

**Comment by bereng** [comment]
CI [lgtm](https://app.circleci.com/pipelines/github/bereng/cassandra/479/workflows/acdc123f-9382-4b1a-8b1b-f58d77df3aa0)

File: src/java/org/apache/cassandra/config/Schema.java [hunk 1]
```diff
@@ -17,16 +17,26 @@
...
 
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/config/Schema.java]: nit: If we're not actually making any functional changes to `Schema`, might as well leave this file alone entirely.

File: test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java [hunk 2]
```diff
@@ -61,6 +66,7 @@
 import org.apache.cassandra.utils.ByteBufferUtil;
 import org.apache.cassandra.utils.FBUtilities;
 import org.apache.cassandra.utils.Pair;
+import org.jboss.byteman.contrib.bmunit.BMRule;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java:72]: @bereng I think we need to use `@RunWith(BMUnitRunner.class)` in this class for the rule to actually be installed. When I tried to run `testNoVisiblePartialSchemaUpdates()` without it, the sleep never happened.

File: test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java [hunk 3]
```diff
@@ -101,8 +107,68 @@ public static void defineSchema() throws ConfigurationException
...
+        method = SchemaKeyspace.class.getDeclaredMethod("mergeSchema", Keyspaces.class, Keyspaces.class);
+        assertTrue(Modifier.isSynchronized(method.getModifiers()));        
+    }
+
+    /** See CASSANDRA-16856/16996. Make sure schema pulls are synchronized to prevent concurrent schema pull/writes */
+    @Test
+    @BMRule(name = "delay partition updates to schema tables",
+            targetClass = "ColumnFamilyStore",
+            targetMethod = "apply",
+            action = "Thread.sleep(5000);",
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java:136]: If we make sure to use `BMUnitRunner.class` (see above), the sleep could probably be lowered again to what it is in 4.0/trunk.

File: test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java [hunk 4]
```diff
@@ -101,8 +107,68 @@ public static void defineSchema() throws ConfigurationException
...
+        CyclicBarrier barrier = new CyclicBarrier(2);
+
+        Future<Void> creation = pool.submit(() -> {
+            barrier.await();
+            createTable(keyspace, "CREATE TABLE test (a text primary key, b int, c int)");
+            return null;
+        });
+
+        Future<Collection<Mutation>> mutationsFromThread = pool.submit(() -> {
+            barrier.await();
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java:153]: The one way this test can still end up not testing what we want it to test is if the create gets delayed entirely by the time this runs.

If you change it to...

```
Future<Collection<Mutation>> mutationsFromThread = pool.submit(() -> {
    barrier.await();

    Collection<Mutation> mutations = SchemaKeyspace.convertSchemaToMutations();
    
    // Make sure we actually have a mutation to check for partial modification.
    while (mutations.size() == 0)
        mutations = SchemaKeyspace.convertSchemaToMutations();
    
    return mutations;
});
```
...it works and doesn't require any new sleeps. I'd probably change 4.0 and trunk to do that as well, just to make sure they actually hit the scenario we want. (Sorry, a bit of an oversight in my original test sketch.)

**Review by maedhroz — CHANGES_REQUESTED** [review]
Dropped a few minor comments, and then there are a couple changes to the test we should probably make sure we get in. Let me know if any of them don't make sense.

File: test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java (hunk 2 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java:72]: Ah I removed it by accident somehow apologies.

File: test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java (hunk 4 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/schema/SchemaKeyspaceTest.java:153]: Which is what I was trying to explain about my test being deterministic and yours not. Either we added the tableId check to mine or added determinism to yours. Now I am happy with this test :-)

**Comment by bereng** [comment]
Nicer test indeed now addressing my concern on it. Pushed changes.

**Comment by bereng** [comment]
CI lgtm

**Comment by bereng** [comment]
CI [lgtm](https://app.circleci.com/pipelines/github/bereng/cassandra/485/workflows/434a2fbb-68f4-4627-a7f5-05e9b7b87f35)

**Pull request closed without merging**